README
DIF - Dynamic Index Facility (DIF) for sky indexing
Giorgio Calderone - giorgio.calderone@inaf.it
Luciano Nicastro - luciano.nicastro@inaf.it

Welcome to DIF.

DIF is free software. Please see the file COPYING for details.  For
documentation, please see dif.pdf. For building and installation instructions
please see the INSTALL file.


ChangeLog
2023-01-28 LN, ver. 0.5.5
	- Various changes to fix "multiple definition" of functions detected by gcc version >= 10

2022-03-29 LN, ver. 0.5.5
	- Fixed integer check in dif.in

2022-01-28 LN, ver. 0.5.5
	- Minor changes to the dif script

2021-12-14 LN, ver. 0.5.5
	- Fixed string offset value in the SQL procedure difview_healp

2021-10-29 LN, ver. 0.5.5
	- Avoid duplicated indexing of unmatched objects table in pix_myXmatch.cc and pix_myXmatch_id.cc

2021-10-29 LN, ver. 0.5.5
	- pix_myXmatch.cc and pix_myXmatch_id.cc bug fixes for full catalogue scan

2021-10-20 LN, ver. 0.5.5
	- Renamed origID to refcatID in the output DB table generated by pix_myXmatch and pix_myXmatch_id

2021-10-20 LN, ver. 0.5.5
	- Added compilation flag -std=c++1z required to compile ha_dif_my8.cc in MySQL v. 8.0.27.

2021-10-14 LN, ver. 0.5.5
	- Updated RAcol, DECcol, RADECcol SQL functions and funcdesc.sql

2021-05-13 LN, ver. 0.5.5
	- commented DBUG_ASSERT in ha_dif.cc because removed in MySQL v. 5.7.34.

2021-04-22 LN, ver. 0.5.5
	- added -DNDEBUG option in configure(.ac) as requested by MySQL v. 8.0.24.

2021-04-21 LN, ver. 0.5.5
	- commented DBUG_ASSERT in ha_dif_my8.cc because removed in MySQL v. 8.0.24.

2020-08-24 LN, ver. 0.5.5
	- Added -fno-rtti for MariaBD only in configure.ac

2020-03-17 LN, ver. 0.5.5
	- Added getRaDec and RADECcol SQL functions.
	- Minor change to configure.ac, RAcol and DECcol.
	- Documentation updated.

2020-03-17 LN, ver. 0.5.5
	- Fixed rectangular region selection defined by 4 corners.
	  Affected file: src/udf.cc
	  Affected DIF function: DIF_RectV
	- README.md documentation updated in main page and doc dir.

2020-01-28 LN, ver. 0.5.5
	- Removed -fno-rtti from ha_dif_la_CXXFLAGS in src/Makefile.am
	- README.md documentation updated

2019-05-07 LN, ver. 0.5.5
	- configure.ac updated in order to use ax_cxx_compile_stdcxx.m4 to request
	  compiling (possibly) using "-std=c++14". This is needed by MySQL v. 8.0.16. 

2019-01-29 LN, ver. 0.5.5
	- configure.ac updated in order to allow installation with MariaDB 10.3.12:
	  libmariadbd.so replaces libmariadb.so
	- Updated: src/ha_dif.h, src/ha_dif.cc, src/udf.cc
	- In HEALPix lib. Makefile uses -std=c++x0 instead of -std=c++11 to allow
	  compilation with gcc 4.4.7 (CentOS 6.10)
	- Removed dif.tex and dif.pdf from doc (Makefile.am updated)
	- Added README.md documentation and png images in doc, doc/includes

2018-11-29 LN, ver. 0.5.5
	- Various changes in order to allow installation with MariaDB 10.3 compiled
	  version.
	- Updated: configure.ac, src/ha_dif.h, src/ha_dif.cc, src/udf.cc,
	  src/ha_dif_my8.cc, src/my_stmt_db.h, src/my_stmt_db2.h
	- Added: src/ha_dif_maria.h

2018-11-16 LN, ver. 0.5.5
	- Various changes in order to allow:
	  1. installation using only the compiled version of MySQL (same version as
	     the system installed version)
	  2. compatibility with MySQL 8
	- Updated: configure.ac configure src/Makefile.am, src/ha_dif.h, src/ha_dif.cc,
	  src/udf.cc, src/my_stmt_db.h, src/my_stmt_db2.h,
	  sql/Makefile.am, sql/funcdesc.sql.in, sql/Messier.sql.in
	- Added: src/ha_dif_my8.cc, sql/func_insert.sql, sql/messier_insert.sql
	- Code cleaning.

2018-10-12 LN, ver. 0.5.4
	- Updated configure.ac to manage cmake build directory.

2018-05-22 LN, ver. 0.5.4
	- Ported to GitHub. Added README.md

2017-10-19 LN, ver. 0.5.4
	- Makefile fix for MariaDB 10.2.9

2017-08-09 LN, ver. 0.5.4
	- Updated ha_dif.cc to support MySQL versions > 5.7 and MariaDB > 10.1

2017-06-08 LN, ver. 0.5.4
	- Fix returned list of partial pixels in DIFmyHealPRect.cpp.

2016-11-09 LN, ver. 0.5.4
	- Added support for MySQL > 5.7.9
	- Use mysql_version.h
	- Additional check for DEC in dif script

2016-08-03 LN, ver. 0.5.4
	- Changed DIF.tbl column definitions.
	- Added RAcol and DECcol SQL procedures.
	- Added sub dif_update in dif script.
	- Added --ra-key option to dif script.

2016-07-11 LN, ver. 0.5.4
	- Added rectangular region selections for HEALPix indexed tables.
	- Added HEALPBound and HEALPBoundC UDFs.
	- Added difview_htmClean, difview_healpClean, difview_Check and difInfo SQL
	  procedures.
	- Added DIF.func table.
	- Added --multidx, --trigger, --cnf, --user, --pass options and --grant-user
	  command to dif script. See dif -h.
	- Code cleaning.

2016-05-16 LN, ver. 0.5.4
	- Replaced HEALPix C++ library with version 3.30

2015-12-23 LN, ver. 0.5.3
	- Works also with MariaDB 10.1

2015-10-17 LN, ver. 0.5.3
	- Added UDF HTMidByName and HTMnameByID.
	- Code cleaning.
	- Documentation updated.

2015-03-07 LN, ver. 0.5.3
	- Added --no-multidx and --no-trigger flags to the dif script.

2015-01-21 LN, ver. 0.5.3
	- Changed coordinates field name management in dif for `Dec` name.
	  Full management is still to be completed.

2014-12-12 LN, ver. 0.5.3
	- Added quotes for coordinates field name (e.g. for `Dec` name).

2014-07-25 LN, ver. 0.5.3
	- Added extra check on input HTM IDs in UDFs and DIF_sNeighb.
	- Documentation updated.

2014-03-30 LN, ver. 0.5.3
	- Works also for MySQL 5.6 and 5.7 (and MySQL Cluster 7.2 and MariaDB 10.0)
	- Added UDF HTMsNeighb and DIF function DIF_sNeighb to get higher
	  depth trixels around a given trixel.
	- Use memset instead of bzero.
	- Added --views-only flag to the dif script.
	- Documentation updated.

2013-01-30 LN, ver. 0.5.2
	- Works also for MySQL 5.5 (and MySQL Cluster 7.2)

2012-08-29 LN, ver. 0.5.2
	- Debian pthread and demo programs building fix.
	- Made DETERMINISTIC stored functions in DIF_udf.sql 

2012-04-10 LN, ver. 0.5.2
	- Mac OS X + MacPorts full support.

2010-09-16 LN, ver. 0.5.2
	- Fixed logging of queries for the "--readonly --log / --logfile"
	  options in dif.

2010-03-15 LN & GC, ver. 0.5.2
	- Comments To Be Added

2010-01-11 LN, ver. 0.5.1
	- Further bugfixes for rectangular region handling functions.
2009-11-24 LN & GC, ver. 0.5.1
	- Bugfix in rectangular region handling function DIF_HTMRect.
	- Bugfix in DIF_Region.
	- Minor fix in dif.
2009-07-27 LN & GC, ver. 0.5.0

src/ha_dif.cc
// ----------------------------------------------------------------------^
// Copyright (C) 2004 - 2018
// Giorgio Calderone <giorgio.calderone@inaf.it>
// Luciano Nicastro <luciano.nicastro@inaf.it>
// 
// This file is part of DIF.
// 
// DIF is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
// 
// DIF is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with DIF; if not, write to the Free Software
// Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
// 
// ----------------------------------------------------------------------$

#ifdef USE_PRAGMA_IMPLEMENTATION
#pragma implementation				// gcc: Class implementation
#endif

#include "dif.hh"

#if MY_VERSION_ID >= 100000

#include "ha_dif_maria.h"

#include <my_config.h>
#include <mysql/plugin.h>
#include "sql_class.h"

#else

#include "ha_dif.h"

// version 5.5.x doesn't contain mysql_priv.h . We need to add the provided includes.
#if MYSQL_VERSION_ID >= 50505

#if MYSQL_VERSION_ID < 50700  // 5.6

#include "sql_priv.h"
#include "sql_class.h"           // MYSQL_HANDLERTON_INTERFACE_VERSION
#include "probes_mysql.h"

#else

// These two are not present in 5.7.9
#if MYSQL_VERSION_ID < 50709
#include <my_pthread.h>
#include <sql_priv.h>
#endif

#endif

//#include <my_global.h>
#include "sql_plugin.h"

#endif

#endif  // >= 100000


#if MY_VERSION_ID >= 50600

static handler *dif_create_handler(handlerton *hton,
                                       TABLE_SHARE *table,
                                       MEM_ROOT *mem_root);

handlerton *dif_hton;

/* Interface to mysqld, to check system tables supported by SE */
static const char* dif_system_database();
static bool dif_is_supported_system_table(const char *db,
                                      const char *table_name,
                                      bool is_sql_layer_system_table);


#ifdef HAVE_PSI_INTERFACE
static PSI_mutex_key ex_key_mutex_Dif_share_mutex;

static PSI_mutex_info all_dif_mutexes[]=
{
  { &ex_key_mutex_Dif_share_mutex, "Dif_share::mutex", 0}
};

static void init_dif_psi_keys()
{
  const char* category= "dif";
  int count;

  count = static_cast<int>(array_elements(all_dif_mutexes));
  mysql_mutex_register(category, all_dif_mutexes, count);
}
#endif


Dif_share::Dif_share()
{
  thr_lock_init(&lock);
  mysql_mutex_init(ex_key_mutex_Dif_share_mutex,
                   &mutex, MY_MUTEX_INIT_FAST);
}


Dif_share *ha_dif::get_share()
{
  Dif_share *tmp_share;

  DBUG_ENTER("ha_dif::get_share()");

  lock_shared_ha_data();
  if (!(tmp_share= static_cast<Dif_share*>(get_ha_share_ptr())))
  {
    tmp_share= new Dif_share;
    if (!tmp_share)
      goto err;

    set_ha_share_ptr(static_cast<Handler_share*>(tmp_share));
  }
err:
  unlock_shared_ha_data();
  DBUG_RETURN(tmp_share);
}

#if MY_VERSION_ID >= 50709  &&  MY_VERSION_ID < 100000
#include <sql/log.h>
#include <sql/auth/auth_common.h>
#endif

#else

#include <mysql/plugin.h>

#endif  // >= 50600


extern ThreadSpecificData<DIF_Region> difreg;


/* Static declarations for handlerton */

static handler *dif_create_handler(handlerton *hton,
                                         TABLE_SHARE *table,
                                         MEM_ROOT *mem_root)
{
  return new (mem_root) ha_dif(hton, table);
}


/*****************************************************************************
** DIF tables
*****************************************************************************/

ha_dif::ha_dif(handlerton *hton,
	       TABLE_SHARE *table_arg)
  :handler(hton, table_arg)
{
    int id_type;
    int id_opt;
    int param;
}


static const char *ha_dif_exts[] = { NullS };
const char **ha_dif::bas_ext() const
{
  return ha_dif_exts;
}

int ha_dif::open(const char *name, int mode, uint test_if_locked)
{
  DBUG_ENTER("ha_dif::open");

//#if defined(MYSQL5_6) || defined(MYSQL5_7) || defined(MYSQL10_0) || defined(MYSQL10_1)
#if ! defined(MYSQL5_1) && ! defined(MYSQL5_5)
  if (!(share = get_share()))
    DBUG_RETURN(1);
  thr_lock_data_init(&share->lock,&lock,NULL);

#else
  thr_lock_init(&thr_lock);
  thr_lock_data_init(&thr_lock,&lock,NULL);
#endif

  //Parse file name
  //const char* p = name;
  //const char* q;
  //while (q = strchr(p, '/')) 
  //    p = q+1;
  //
  //if (strncmp(p, "htm", 3) == 0)
  //    id_type = 1;
  //
  //if (strncmp(p, "hea", 3) == 0)
  //    id_type = 2;
  //
  //char buf[2];
  //buf[0] = *(p+4);
  //buf[1] = 0;
  //sscanf(buf, "%d", &id_opt);
  //sscanf(p+6, "%d", &param);

  DBUG_RETURN(0);
}


int ha_dif::close(void)
{
  DBUG_ENTER("ha_dif::close");

//#if ! defined(MYSQL5_5) && ! defined(MYSQL5_6) && ! defined(MYSQL5_7) && ! defined(MYSQL10_0) && ! defined(MYSQL10_1)
#if defined(MYSQL5_1) || defined(MYSQL5_5)
  thr_lock_delete(&thr_lock);
#endif

  DBUG_RETURN(0);
}


int ha_dif::create(const char *name, TABLE *table_arg,
                         HA_CREATE_INFO *create_info)
{
  DBUG_ENTER("ha_dif::create");
  DBUG_RETURN(0);
}

/*
const char *ha_dif::index_type(uint key_number)
{
  DBUG_ENTER("ha_dif::index_type");
  DBUG_RETURN((table_share->key_info[key_number].flags & HA_FULLTEXT) ?
              "FULLTEXT" :
              (table_share->key_info[key_number].flags & HA_SPATIAL) ?
              "SPATIAL" :
              (table_share->key_info[key_number].algorithm ==
               HA_KEY_ALG_RTREE) ? "RTREE" : "BTREE");
}
*/

int ha_dif::write_row(uchar * buf)
{
  DBUG_ENTER("ha_dif::write_row");
  DBUG_RETURN(0);
}


int ha_dif::rnd_init(bool scan)
{
  DBUG_ENTER("ha_dif::rnd_init");

  if (! difreg.getp()) difreg.constructor();
  difreg->subStart();
  difreg->go();
  difreg->subStop();

  DBUG_RETURN(0);
}

//Added

int ha_dif::rnd_end() {
  DBUG_ENTER("ha_dif::rnd_end");
  DBUG_RETURN(0);
}


int ha_dif::rnd_next(uchar *buf)
{
  DBUG_ENTER("ha_dif::rnd_next");

  int param, full;
  long long int val;

  difreg->subStart();
  int eof = difreg->read_next(param, val, full);

  if (! eof) {
      Field **field=table->field;
      (*field)->set_notnull();  field++;
      (*field)->set_notnull();  field++;
      (*field)->set_notnull();
      
      field=table->field;
      (*field)->store(param);   field++;
      (*field)->store(val);     field++;
      (*field)->store(full);
      difreg->subStop();
      DBUG_RETURN(0);
  }

  difreg->read_reset();

  //if (eof == 1) 
  //  difreg->clear();  //Final EOF
  difreg->subStop();

  DBUG_RETURN(HA_ERR_END_OF_FILE);
}


int ha_dif::rnd_pos(uchar * buf, uchar *pos)
{
  DBUG_ENTER("ha_dif::rnd_pos");
  //  5.7.34 removed DBUG_ASSERT(0);
  assert(0);
  DBUG_RETURN(0);
}


void ha_dif::position(const uchar *record)
{
  DBUG_ENTER("ha_dif::position");
  // 5.7.34 removed DBUG_ASSERT(0);
  assert(0);
  DBUG_VOID_RETURN;
}


#ifdef VOID_HANDLER_INFO
  void ha_dif::info(uint flag)
#else
  int ha_dif::info(uint flag)
#endif
{
  DBUG_ENTER("ha_dif::info");

  memset((char*) &stats, 0, sizeof(stats));
  if (flag & HA_STATUS_AUTO)
    stats.auto_increment_value= 1;

#ifndef VOID_HANDLER_INFO
  DBUG_RETURN(0);
#endif
}


int ha_dif::external_lock(THD *thd, int lock_type)
{
  DBUG_ENTER("ha_dif::external_lock");
  DBUG_RETURN(0);
}

/*
uint ha_dif::lock_count(void) const
{
  DBUG_ENTER("ha_dif::lock_count");
  DBUG_RETURN(0);
}
*/

THR_LOCK_DATA **ha_dif::store_lock(THD *thd,
                                         THR_LOCK_DATA **to,
                                         enum thr_lock_type lock_type)
{
  DBUG_ENTER("ha_dif::store_lock");
  DBUG_RETURN(to);
}

/*
int ha_dif::index_read(uchar * buf, const uchar * key,
                             uint key_len, enum ha_rkey_function find_flag)
{
  DBUG_ENTER("ha_dif::index_read");
  DBUG_RETURN(0);
}


int ha_dif::index_read_idx(uchar * buf, uint idx, const uchar * key,
                                 uint key_len, enum ha_rkey_function find_flag)
{
  DBUG_ENTER("ha_dif::index_read_idx");
  DBUG_RETURN(HA_ERR_END_OF_FILE);
}


int ha_dif::index_read_last(uchar * buf, const uchar * key, uint key_len)
{
  DBUG_ENTER("ha_dif::index_read_last");
  DBUG_RETURN(HA_ERR_END_OF_FILE);
}
*/

int ha_dif::index_next(uchar * buf)
{
  DBUG_ENTER("ha_dif::index_next");
  DBUG_RETURN(HA_ERR_END_OF_FILE);
}


int ha_dif::index_prev(uchar * buf)
{
  DBUG_ENTER("ha_dif::index_prev");
  DBUG_RETURN(HA_ERR_END_OF_FILE);
}


int ha_dif::index_first(uchar * buf)
{
  DBUG_ENTER("ha_dif::index_first");
  DBUG_RETURN(HA_ERR_END_OF_FILE);
}


int ha_dif::index_last(uchar * buf)
{
  DBUG_ENTER("ha_dif::index_last");
  DBUG_RETURN(HA_ERR_END_OF_FILE);
}


static int dif_init(void *p)
{
  handlerton *dif_hton;
  dif_hton= (handlerton *)p;
  dif_hton->state= SHOW_OPTION_YES;
  dif_hton->db_type= DB_TYPE_UNKNOWN;
  dif_hton->create= dif_create_handler;
  dif_hton->flags= HTON_CAN_RECREATE;
  return 0;
}


struct st_mysql_storage_engine dif_storage_engine=
{ MYSQL_HANDLERTON_INTERFACE_VERSION };


#if MY_VERSION_ID >= 100000
maria_declare_plugin(dif)
{
  MYSQL_STORAGE_ENGINE_PLUGIN,
  &dif_storage_engine,
  "DIF",
  "Giorgio Calderone & Luciano Nicastro, INAF (Italy)",
  "Dynamic Index Facility",
  PLUGIN_LICENSE_GPL,
  dif_init, /* Plugin Init */
  NULL, /* Plugin Deinit */
  0x0100 /* 1.0 */,
  NULL,                       /* status variables                */
  NULL,                       /* system variables                */
  "1.0",                      /* string version */
  MariaDB_PLUGIN_MATURITY_STABLE  /* maturity */
}
maria_declare_plugin_end;

#else

mysql_declare_plugin(dif)
{
  MYSQL_STORAGE_ENGINE_PLUGIN,
  &dif_storage_engine,
  "DIF",
  "Giorgio Calderone & Luciano Nicastro, INAF (Italy)",
  "Dynamic Index Facility",
  PLUGIN_LICENSE_GPL,
  dif_init, /* Plugin Init */
  NULL, /* Plugin Deinit */
  0x0100 /* 1.0 */,
  NULL,                       /* status variables                */
  NULL,                       /* system variables                */
  NULL                        /* config options                  */
}
mysql_declare_plugin_end;

#endif
